File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.cpp
#include <string>
#include <vector>
#include <queue>
#include <unordered_map> 
#include <sstream>
#include <iostream>
using namespace std;
/*
명령어에 따라 작업 수행
비어있는데 삭제하라고 하면 무시
최종 결과가 비어있다면 [0,0] 비어잇지 않다면 [최대값, 최솟값]
*/
vector<int> solution(vector<string> operations) {
    vector<int> answer;
    
    priority_queue<int> max_heap; 
    priority_queue<int, vector<int>, greater<int>> min_heap; 
    
    unordered_map<int, int> m; // 상태 파악할 수 있는 map {숫쟈, 남은 개수}
    
    for(int i = 0; i < operations.size(); i++){
        string order = operations[i];
        stringstream ss(order); // 공백 기준으로 분리
        string o; ss >> o; // 명령어
        string ns; ss >> ns; // "-1" 이렇게 하나로 받아야 함
        
        if(o == "I"){ // 삽입
            int n = stoi(ns);
            max_heap.push(n); min_heap.push(n);
            m[n]++;
            //cout << "insert : " << n << " " << m[n] << endl;
        }
        else{ // 삭제
            if(ns == "-1"){ // 최솟값 삭제
                bool flag = false;
                while(!min_heap.empty()){
                    int num = min_heap.top(); 
                    if(m.find(num) == m.end()){
                        min_heap.pop();
                    } else {
                        flag = true; break;
                    }
                }
                if(flag){
                    int num = min_heap.top(); 
                    //cout << "del: " << num << " " << m[num] << endl;
                    if(--m[num] == 0) m.erase(num); 
                }

                
            }
            else{ // 최댓값 삭제
                bool flag = false;
                while(!max_heap.empty()){
                    int num = max_heap.top(); 
                    if(m.find(num) == m.end()){
                        max_heap.pop();
                    } else {
                        flag = true; break;
                    }
                }
                if(flag){
                    int num = max_heap.top(); 
                    //cout << "del: " << num << " " << m[num] << endl;
                    if(--m[num] == 0) m.erase(num); 
                }
                
            }
            
        }
        
    }
    
    
    // 최댓값, 최솟값 찾아 넣기
    if(m.size() == 0) answer = {0, 0}; 
    else{
        while(m.size() != 0){
            int num = max_heap.top(); 
            if(m.find(num) == m.end()){
                max_heap.pop();
            } else break;
        }
        int maxx = max_heap.top();
        
        while(m.size() != 0){
            int num = min_heap.top(); 
            if(m.find(num) == m.end()){
                min_heap.pop();
            } else break;
        }
        int minn = min_heap.top(); 
        
        answer = {maxx, minn};
        
    }
    
    
    return answer;
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*; 

class Solution {
    public int[] solution(String[] operations) { // 1,000,000 
        
        HashMap<Integer, Integer> nums = new HashMap<>(); 
        
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> {
            return Integer.compare(b, a);
        }); 
        PriorityQueue<Integer> minHeap = new PriorityQueue<>((a, b) -> {
            return Integer.compare(a, b);
        }); 
        
        for(String operation : operations){
            String[] order = operation.split(" ");
            if(order[0].equals("I")){ // 삽입
                int n = Integer.parseInt(order[1]); 
                // System.out.println("insert: " + n);
                nums.merge(n, 1, Integer::sum);
                minHeap.offer(n); maxHeap.offer(n);
                
            }else if(order[0].equals("D")){
                if(order[1].equals("1")){ // 최댓값 삭제              
                    deleteMax(maxHeap,nums);
                }else{ // 최솟값 삭제
                    deleteMin(minHeap, nums);

                }
            }
        }
        
        // 모든 연산을 처리한 후 큐가 비어있으면 [0,0] 비어있지 않으면 [최댓값, 최솟값]을 return
        int min = 0; int max = 0;
        
        String m = deleteMax(maxHeap, nums); 
        if(!m.equals("-")) max = Integer.parseInt(m);
        minHeap.offer(max); nums.merge(max, 1, Integer::sum);
        
        m = deleteMin(minHeap, nums); 
        if(!m.equals("-")) min = Integer.parseInt(m);
        
        return new int[]{max, min};
    }
    
    public String deleteMax(PriorityQueue<Integer> maxHeap, HashMap<Integer, Integer> nums){
        int max = 0; boolean avail = false;
        while(maxHeap.size() != 0) { 
            // 이제 heap에 데이터가 없음
            if(maxHeap.size() == 0) break;
            // 데이터 뽑기
            max = maxHeap.poll();
            // 데이터가 이미 뽑혔음
            if(!nums.containsKey(max) || (nums.containsKey(max) && nums.get(max) <= 0)) continue;
            // 뽑혔으니 데이터 제거
            nums.merge(max, -1, Integer::sum);
            // System.out.println("del: " + max);
            avail = true;
            break;
        }
        if(avail) return String.valueOf(max);
        else return "-";
        
    }
    public String deleteMin(PriorityQueue<Integer> minHeap, HashMap<Integer, Integer> nums){
        int min = 0; boolean avail = false;
        while(minHeap.size() != 0) { 
            // 이제 heap에 데이터가 없음
            if(minHeap.size() == 0) break;
            // 데이터 뽑기
            min = minHeap.poll();
            // 데이터가 이미 뽑혔음
            if(!nums.containsKey(min) || (nums.containsKey(min) && nums.get(min) <= 0)) continue;
            // 뽑혔으니 데이터 제거
            nums.merge(min, -1, Integer::sum);
            // System.out.println("del: " + min);
            avail = true;
            break;
        }
        if(avail) return String.valueOf(min);
        else return "-";
    }
}